src/charm.py

```diff
@@ -215,21 +216,63 @@ def _restart(self, event: EventBase) -> None:
-        if self.app_peer_data.get("rotate-passwords"):
-            self.unit_peer_data["password-rotated"] = "true"
-
-        self.unit_peer_data.update(
+        self.state.unit_server.update(
             {
                 # flag to declare unit running `portUnification` during ssl<->no-ssl upgrade
-                "unified": "true" if self.tls.upgrading else "",
+                "unified": "true" if self.state.cluster.switching_encryption else "",
                 # flag to declare unit restarted with new quorum encryption
-                "quorum": self.cluster.quorum,
+                "quorum": self.state.cluster.quorum,
+                # indicate that unit has completed restart on password rotation
+                "password-rotated": "true" if self.state.cluster.rotate_passwords else "",
             }
         )
 
-        if self.provider.ready:
-            self.provider.apply_relation_data()
+        self.update_client_data()
+
+    def _get_super_password_action(self, event: ActionEvent) -> None:
+        """Handler for get-super-password action event."""
+        event.set_results(
+            {"super-password": self.state.cluster.internal_user_credentials["super"]}
+        )
+
+    def _get_sync_password_action(self, event: ActionEvent) -> None:
+        """Handler for get-sync-password action event."""
+        event.set_results({"sync-password": self.state.cluster.internal_user_credentials["sync"]})
+
+    def _set_password_action(self, event: ActionEvent) -> None:
+        """Handler for set-password action.
+
+        Set the password for a specific user, if no passwords are passed, generate them.
+        """
+        if not self.unit.is_leader():
+            msg = "Password rotation must be called on leader unit"
+            logger.error(msg)
+            event.fail(msg)
+            return
+
+        username = event.params.get("username", "super")
+        if username not in CHARM_USERS:
+            msg = f"The action can be run only for users used by the charm: {CHARM_USERS} not {username}."
+            logger.error(msg)
+            event.fail(msg)
+            return
+
+        new_password = event.params.get("password", generate_password())
+
+        # Passwords should not be the same.
+        if new_password in self.state.cluster.internal_user_credentials.values():
+            event.log("The old and new passwords are equal.")
+            event.set_results({f"{username}-password": new_password})
+            return
+
+        # Store those passwords on application databag
+        self.state.cluster.update({f"{username}-password": new_password})
+
+        # implicitly calls config_changed on leader, other units will get it because of relation-data change with new passwords
+        self._on_cluster_relation_changed(event)
+        event.set_results({f"{username}-password": new_password})
+
+    # --- CONVENIENCE METHODS ---
 
     def init_server(self):
```

Comment:
Mostly, yeah!
There are some minor differences I think, but it's mostly the same.

The reasons that it's in a separate function (as opposed to event-handlers only) are:
1. It makes it easier to mock out in scenarios where we don't actually care about properly initialising a running ZK service
2. It's very long and very complicated, and clutters the already bloated `cluster-relation-changed` handler

The same reasoning applies to `update_quorum` below.

More practically, it's the part of the charm that requires the most debugging, and changes the most when adding features that might mess with the quorum. It helps having it disconnected from the actual event handling, otherwise it's a huge complexity headache with all the conditional paths.